lms/services/assignment.py

```diff
@@ -56,61 +91,58 @@ def get_copied_from_assignment(self, lti_params) -> Optional[Assignment]:
-        historical assignment (perhaps by using the LMS's "course copy" feature)
-        then some of the new assignment's extra values might be taken from the historical assignment.
+        Returns None if no assignment can be found or created.
         """
-        extra = extra or {}
 
+        lti_params = request.lti_params
+        tool_consumer_instance_guid = lti_params["tool_consumer_instance_guid"]
+        resource_link_id = lti_params.get("resource_link_id")
+
+        # Get the potentially relevant assignments from the DB
         assignment = self.get_assignment(tool_consumer_instance_guid, resource_link_id)
+        historical_assignment = None
         if not assignment:
-            assignment = Assignment(
-                tool_consumer_instance_guid=tool_consumer_instance_guid,
-                document_url=document_url,
-                resource_link_id=resource_link_id,
-            )
-            self._db.add(assignment)
+            historical_assignment = self._get_copied_from_assignment(lti_params)
 
-            # For new assignments check if we are copying
-            # from an existing one on the current launch
-            if historical_assignment := self.get_copied_from_assignment(lti_params):
-                assignment.copied_from = historical_assignment
+        # Get the configuration for the assignment
+        # it might be based on the assignments we just queried or the request
+        document_url = self._misc_plugin.get_document_url(
+            request, assignment, historical_assignment
+        )
+        group_set_id = self._grouping_plugin.get_group_set_id(
+            request, assignment, historical_assignment
+        )
 
-                if historical_assignment.extra.get("group_set_id") and not extra.get(
-                    "group_set_id"
-                ):
-                    extra["group_set_id"] = historical_assignment.extra.get(
-                        "group_set_id"
-                    )
+        if not document_url:
+            # We can't find a document_url, we shouldn't try to create an assignment yet.
+            return None
 
-        assignment.document_url = document_url
-        assignment.title = lti_params.get("resource_link_title")
-        assignment.description = lti_params.get("resource_link_description")
-        assignment.is_gradable = is_gradable
-        if extra:
-            assignment.extra = extra
+        if not assignment:
+            # We don't have an assignment in the DB but we know to which document url it should point
+            # This might happen for example on:
+            #   - the first launch of a deep linked assignment
+            #   - the first launch copied assignment
+            assignment = self.create_assignment(
+                tool_consumer_instance_guid, resource_link_id
+            )
 
-        return assignment
+            if historical_assignment:
```

Comment:
Do we need to if here? If historical assignment is None, presumably it would have no effect?